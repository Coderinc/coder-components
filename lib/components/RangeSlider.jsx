import React from 'react';
import PropTypes from 'prop-types';
import RCSlider from 'rc-slider';
import { withTheme } from 'styled-components';

const rangeSliderStyle = theme => ({
  railStyle: {
    backgroundColor: theme.offsetBackground,
  },
  dotStyle: {
    display: 'none',
  },
  activeDotStyle: {
    backgroundColor: theme.primary,
    borderColor: theme.primaryBackground,
    borderWidth: 2,
  },
  trackStyle: [
    {
      backgroundColor: theme.primary,
    },
  ],
  handleStyle: [
    {
      backgroundColor: theme.primary,
      borderColor: theme.primaryBackground,
    },
    {
      backgroundColor: theme.primary,
      borderColor: theme.primaryBackground,
    },
  ],
});

const markStyle = (theme, index, active) => {
  if (index !== 0 && index !== 10 && !active) {
    return { display: 'none' };
  }
  return {
    color: active ? theme.primaryFont : theme.offsetFont,
  };
};

class RangeSlider extends React.Component {
  constructor(props) {
    super(props);

    const { minValue = 0, maxValue = 10 } = props;
    this.marks = {};
    const delta = Math.round((maxValue - minValue) / 10);
    for (let i = 0; i < 11; i += 1) {
      const val = minValue + i * delta;
      this.marks[val] = {
        label: `${val}`,
        style: markStyle(props.theme, i, false),
      };
    }
  }

  render() {
    const {
      minValue, maxValue, input, theme,
    } = this.props;

    const marks = {};
    Object.keys(this.marks).forEach((k, i) => {
      marks[k] = {
        ...this.marks[k],
        style: markStyle(theme, i, input.value.includes(+k)),
      };
    });

    return (
      <RCSlider.Range
        defaultValue={input.value || []}
        marks={marks}
        max={maxValue}
        min={minValue}
        onChange={input.onChange}
        step={null}
        style={{ margin: '1rem 0' }}
        value={input.value}
        {...rangeSliderStyle(theme)}
      />
    );
  }
}

RangeSlider.propTypes = {
  minValue: PropTypes.number,
  maxValue: PropTypes.number,
  input: PropTypes.object.isRequired,
  theme: PropTypes.object.isRequired,
};

RangeSlider.defaultProps = {
  minValue: 0,
  maxValue: 10,
};

export default withTheme(RangeSlider);
