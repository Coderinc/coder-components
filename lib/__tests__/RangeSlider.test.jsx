import React from 'react';
import renderer from 'react-test-renderer';
import RangeSlider from '../components/RangeSlider';

describe('RangeSlider', () => {
  it('renders correctly', () => {
    const rangeSliderInput = {
      onChange: undefined,
      value: [0, 5],
    };
    const component = renderer.create(<RangeSlider input={rangeSliderInput} theme={{}} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
