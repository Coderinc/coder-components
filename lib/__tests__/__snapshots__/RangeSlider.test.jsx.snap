// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider renders correctly 1`] = `
<div
  className="rc-slider rc-slider-with-marks"
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  style={
    Object {
      "margin": "1rem 0",
    }
  }
>
  <div
    className="rc-slider-rail"
    style={
      Object {
        "backgroundColor": undefined,
      }
    }
  />
  <div
    className="rc-slider-track rc-slider-track-1"
    style={
      Object {
        "backgroundColor": undefined,
        "left": "0%",
        "width": "50%",
      }
    }
  />
  <div
    className="rc-slider-step"
  >
    <span
      className="rc-slider-dot rc-slider-dot-active"
      style={
        Object {
          "backgroundColor": undefined,
          "borderColor": undefined,
          "borderWidth": 2,
          "display": "none",
          "left": "0%",
        }
      }
    />
    <span
      className="rc-slider-dot rc-slider-dot-active"
      style={
        Object {
          "backgroundColor": undefined,
          "borderColor": undefined,
          "borderWidth": 2,
          "display": "none",
          "left": "10%",
        }
      }
    />
    <span
      className="rc-slider-dot rc-slider-dot-active"
      style={
        Object {
          "backgroundColor": undefined,
          "borderColor": undefined,
          "borderWidth": 2,
          "display": "none",
          "left": "20%",
        }
      }
    />
    <span
      className="rc-slider-dot rc-slider-dot-active"
      style={
        Object {
          "backgroundColor": undefined,
          "borderColor": undefined,
          "borderWidth": 2,
          "display": "none",
          "left": "30%",
        }
      }
    />
    <span
      className="rc-slider-dot rc-slider-dot-active"
      style={
        Object {
          "backgroundColor": undefined,
          "borderColor": undefined,
          "borderWidth": 2,
          "display": "none",
          "left": "40%",
        }
      }
    />
    <span
      className="rc-slider-dot rc-slider-dot-active"
      style={
        Object {
          "backgroundColor": undefined,
          "borderColor": undefined,
          "borderWidth": 2,
          "display": "none",
          "left": "50%",
        }
      }
    />
    <span
      className="rc-slider-dot"
      style={
        Object {
          "display": "none",
          "left": "60%",
        }
      }
    />
    <span
      className="rc-slider-dot"
      style={
        Object {
          "display": "none",
          "left": "70%",
        }
      }
    />
    <span
      className="rc-slider-dot"
      style={
        Object {
          "display": "none",
          "left": "80%",
        }
      }
    />
    <span
      className="rc-slider-dot"
      style={
        Object {
          "display": "none",
          "left": "90%",
        }
      }
    />
    <span
      className="rc-slider-dot"
      style={
        Object {
          "display": "none",
          "left": "100%",
        }
      }
    />
  </div>
  <div
    aria-disabled={false}
    aria-valuemax={10}
    aria-valuemin={0}
    aria-valuenow={0}
    className="rc-slider-handle rc-slider-handle-1"
    role="slider"
    style={
      Object {
        "backgroundColor": undefined,
        "borderColor": undefined,
        "left": "0%",
      }
    }
    tabIndex={0}
  />
  <div
    aria-disabled={false}
    aria-valuemax={10}
    aria-valuemin={0}
    aria-valuenow={5}
    className="rc-slider-handle rc-slider-handle-2"
    role="slider"
    style={
      Object {
        "backgroundColor": undefined,
        "borderColor": undefined,
        "left": "50%",
      }
    }
    tabIndex={0}
  />
  <div
    className="rc-slider-mark"
  >
    <span
      className="rc-slider-mark-text rc-slider-mark-text-active"
      style={
        Object {
          "color": undefined,
          "left": "0%",
          "marginLeft": "-4.5%",
          "width": "9%",
        }
      }
    >
      0
    </span>
    <span
      className="rc-slider-mark-text rc-slider-mark-text-active"
      style={
        Object {
          "display": "none",
          "left": "10%",
          "marginLeft": "-4.5%",
          "width": "9%",
        }
      }
    >
      1
    </span>
    <span
      className="rc-slider-mark-text rc-slider-mark-text-active"
      style={
        Object {
          "display": "none",
          "left": "20%",
          "marginLeft": "-4.5%",
          "width": "9%",
        }
      }
    >
      2
    </span>
    <span
      className="rc-slider-mark-text rc-slider-mark-text-active"
      style={
        Object {
          "display": "none",
          "left": "30%",
          "marginLeft": "-4.5%",
          "width": "9%",
        }
      }
    >
      3
    </span>
    <span
      className="rc-slider-mark-text rc-slider-mark-text-active"
      style={
        Object {
          "display": "none",
          "left": "40%",
          "marginLeft": "-4.5%",
          "width": "9%",
        }
      }
    >
      4
    </span>
    <span
      className="rc-slider-mark-text rc-slider-mark-text-active"
      style={
        Object {
          "color": undefined,
          "left": "50%",
          "marginLeft": "-4.5%",
          "width": "9%",
        }
      }
    >
      5
    </span>
    <span
      className="rc-slider-mark-text"
      style={
        Object {
          "display": "none",
          "left": "60%",
          "marginLeft": "-4.5%",
          "width": "9%",
        }
      }
    >
      6
    </span>
    <span
      className="rc-slider-mark-text"
      style={
        Object {
          "display": "none",
          "left": "70%",
          "marginLeft": "-4.5%",
          "width": "9%",
        }
      }
    >
      7
    </span>
    <span
      className="rc-slider-mark-text"
      style={
        Object {
          "display": "none",
          "left": "80%",
          "marginLeft": "-4.5%",
          "width": "9%",
        }
      }
    >
      8
    </span>
    <span
      className="rc-slider-mark-text"
      style={
        Object {
          "display": "none",
          "left": "90%",
          "marginLeft": "-4.5%",
          "width": "9%",
        }
      }
    >
      9
    </span>
    <span
      className="rc-slider-mark-text"
      style={
        Object {
          "color": undefined,
          "left": "100%",
          "marginLeft": "-4.5%",
          "width": "9%",
        }
      }
    >
      10
    </span>
  </div>
</div>
`;
